### DISCUSSION TITLE
Auto-refactor on file renaming?

### DISCUSSION THREAD

**Discussion by dxu** [nvim-mini/mini.nvim#487] [Q&A]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)

### Module(s)

nvim-tree, nvim-lsp-file-operations, 

### Description

Hi, I love your work! I use several of the plugins in my own setup, and absolutely love mini.files. Is there a way to enable refactoring of imports on file renaming, similar to [nvim-lsp-file-operations](https://github.com/antosha417/nvim-lsp-file-operations) for nvim-tree/neotree? 

I've tried to mimic the functionality using the mini.files events, but it doesn't seem to work - but I'm a bit of a neovim noob, so not sure if I'm doing something wrong. 

[Here](https://github.com/dxu/dotvim/blob/master/init.lua#L886-L914) I'm setting up a listener on the `MiniFilesActionRename` event, and trying to mimic what [nvim-lsp-file-operations does for handling `willRenameFiles`](https://github.com/antosha417/nvim-lsp-file-operations/blob/master/lua/lsp-file-operations/will-rename.lua), but it doesn't seem to do anything

**Comment by echasnovski** [comment]
Hi! Thanks!

This kind of action won't be part of 'mini.files' itself, because, yes, utilizing events is the proposed approach here. So I'll convert it to discussion.

**Comment by echasnovski** [comment]
As per [your code](https://github.com/dxu/dotvim/blob/master/init.lua#L886-L914), one obvious thing that stands out is that callbacks receive extra info inside `data` field of input argument. So it should use `data.data.from` and `data.data.to` (instead of `data.from` and `data.to`).

Other than that, I am not sure how 'lsp-file-operations' is supposed to work and if the other code is correct.

**Comment by andvarfolomeev** [comment]
I sent [draft PR](https://github.com/antosha417/nvim-lsp-file-operations/pull/29) to this repo. @dxu, It works but doesn't as I expect. Maybe you can help me to fix it.

**Comment by shubham-cpp** [comment]
@echasnovski , the code provided by @andvarfolomeev doesn't seem to work now.
I've created an issue - https://github.com/antosha417/nvim-lsp-file-operations/issues/37

**Reply by echasnovski** [reply]
I am not sure if this is related to the changes in 'mini.files'. I don't see any relevant changes in 'mini.files' since 2024-05-11 (when it was confirmed to "work fine").

**Reply by shubham-cpp** [reply]
Got it some-what working
```lua
local au_group = vim.api.nvim_create_augroup('__mini', { clear = true })
local events = {
  ['lsp-file-operations.did-rename'] = { { 'MiniFilesActionRename', 'MiniFilesActionMove' }, 'Renamed' },
  ['lsp-file-operations.will-create'] = { 'MiniFilesActionCreate', 'Create' },
  ['lsp-file-operations.will-delete'] = { 'MiniFilesActionDelete', 'Delete' },
}
for module, pattern in pairs(events) do
  vim.api.nvim_create_autocmd('User', {
    pattern = pattern[1],
    group = au_group,
    desc = string.format('Auto-refactor LSP file %s', pattern[2]),
    callback = function(event)
      local ok, action = pcall(require, module)
      if not ok then
        return
      end
      local args = {}
      local data = event.data
      if data.from == nil or data.to == nil then
        args = { fname = data.from or data.to }
      else
        args = { old_name = data.from, new_name = data.to }
      end
      action.callback(args)
    end,
  })
end
```

<details>

<summary>Full config</summary>

```lua
local au_group = vim.api.nvim_create_augroup('__mini', { clear = true })
local minifiles_toggle = function(...)
  local f = require 'mini.files'
  local arg_len = ... == nil and 0 or #...
  if not f.close() then
    if arg_len == 0 then
      f.open(f.get_latest_path())
    else
      f.open(...)
    end
  end
end
---@type LazySpec
return {
  'echasnovski/mini.files',
  dependencies = {
    'nvim-tree/nvim-web-devicons',
    { 'antosha417/nvim-lsp-file-operations', dependencies = { 'nvim-lua/plenary.nvim' } },
  },
  version = '*',
  enabled = true,
  keys = {
      {
        '<leader>e',
        minifiles_toggle,
        desc = 'Toggle file explorer',
      },
      {
        '<leader>E',
        function()
          minifiles_toggle(vim.api.nvim_buf_get_name(0), false)
        end,
        desc = 'Toggle file explorer(current file)',
      },
    },
  config = function()
    require('mini.files').setup({
      options = { permanent_delete = false, use_as_default_explorer = true },
      windows = { preview = true, width_preview = 45 },
    })
    require('lsp-file-operations').setup()

    local events = {
      ['lsp-file-operations.did-rename'] = { { 'MiniFilesActionRename', 'MiniFilesActionMove' }, 'Renamed' },
      ['lsp-file-operations.will-create'] = { 'MiniFilesActionCreate', 'Create' },
      ['lsp-file-operations.will-delete'] = { 'MiniFilesActionDelete', 'Delete' },
    }
    for module, pattern in pairs(events) do
      vim.api.nvim_create_autocmd('User', {
        pattern = pattern[1],
        group = au_group,
        desc = string.format('Auto-refactor LSP file %s', pattern[2]),
        callback = function(event)
          local ok, action = pcall(require, module)
          if not ok then
            return
          end
          local args = {}
          local data = event.data
          if data.from == nil or data.to == nil then -- when the `pattern` is "create" or "delete"
            args = { fname = data.from or data.to }
          else
            args = { old_name = data.from, new_name = data.to }
          end
          action.callback(args)
        end,
      })
    end
  end,
}
```

</details>

**Reply by stefanwatt** [reply]
what do you mean by somewhat? 😄
